File: scripts/validate_manifest.js
#!/usr/bin/env node

const fs = require("node:fs");
const path = require("node:path");
const { isMap, parseDocument } = require("yaml");

const repoRoot = path.resolve(__dirname, "..");
const manifestPath = path.resolve(repoRoot, process.argv[2] ?? "manifest/workstation-state.yaml");
const errors = [];

if (!fs.statSync(manifestPath, { throwIfNoEntry: false })?.isFile()) {
  console.error(`Manifest not found: ${manifestPath}`);
  process.exit(1);
}

const manifestDocument = parseDocument(fs.readFileSync(manifestPath, "utf8"), {
  maxAliasCount: 0
});

if (manifestDocument.errors.length > 0) {
  console.error(
    `Invalid YAML in ${path.relative(repoRoot, manifestPath)}: ${manifestDocument.errors[0].message}`
  );
  process.exit(1);
}

if (!isMap(manifestDocument.contents)) {
  console.error("Manifest root must be a mapping");
  process.exit(1);
}

const manifest = manifestDocument.toJS();
let declaredAdapters = manifest.adapters ?? {};

if (!isPlainObject(declaredAdapters)) {
  errors.push("Top-level 'adapters' must be a mapping");
  declaredAdapters = {};
}

const referencedAdapters = [];
visit(excludeKey(manifest, "adapters"), referencedAdapters);

for (const adapter of [...new Set(referencedAdapters)].sort()) {
  if (!Object.hasOwn(declaredAdapters, adapter)) {
    errors.push(`Referenced adapter '${adapter}' is not declared under top-level 'adapters'`);
  }
}

const contractReference = manifest.metadata?.contract;

if (typeof contractReference !== "string" || contractReference.length === 0) {
  errors.push("metadata.contract must be a non-empty local path");
} else {
  const contractPath = path.resolve(path.dirname(manifestPath), contractReference);
  const relativeContract = path.relative(repoRoot, contractPath);

  if (relativeContract === ".." || relativeContract.startsWith(`..${path.sep}`) || path.isAbsolute(relativeContract)) {
    errors.push("metadata.contract must resolve inside the repository");
  } else if (!fs.statSync(contractPath, { throwIfNoEntry: false })?.isFile()) {
    errors.push(`metadata.contract does not exist: ${relativeContract}`);
  }
}

if (errors.length > 0) {
  errors.forEach((error) => console.error(`ERROR: ${error}`));
  process.exit(1);
}

console.log(`Validated ${path.relative(repoRoot, manifestPath)}`);
console.log(
  `${Object.keys(declaredAdapters).length} adapters declared; ${new Set(referencedAdapters).size} referenced`
);

function excludeKey(object, key) {
  return Object.fromEntries(Object.entries(object).filter(([entryKey]) => entryKey !== key));
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function visit(value, referencedAdapters) {
  if (Array.isArray(value)) {
    value.forEach((child) => visit(child, referencedAdapters));
    return;
  }

  if (!isPlainObject(value)) {
    return;
  }

  for (const [key, child] of Object.entries(value)) {
    if (key === "adapter" && typeof child === "string") {
      referencedAdapters.push(child);
    }

    visit(child, referencedAdapters);
  }
}

File: scripts/validate_manifest.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "pathname"
require "yaml"

repo_root = Pathname(__dir__).parent.realpath
manifest_path = Pathname(ARGV.fetch(0, "manifest/workstation-state.yaml")).expand_path(repo_root)
errors = []

unless manifest_path.file?
  warn "Manifest not found: #{manifest_path}"
  exit 1
end

begin
  manifest = YAML.safe_load(manifest_path.read, aliases: false)
rescue Psych::Exception => e
  warn "Invalid YAML in #{manifest_path.relative_path_from(repo_root)}: #{e.message}"
  exit 1
end

unless manifest.is_a?(Hash)
  warn "Manifest root must be a mapping"
  exit 1
end

declared_adapters = manifest.fetch("adapters", {})
unless declared_adapters.is_a?(Hash)
  errors << "Top-level 'adapters' must be a mapping"
  declared_adapters = {}
end

referenced_adapters = []
visit = lambda do |value|
  case value
  when Hash
    value.each do |key, child|
      referenced_adapters << child if key == "adapter" && child.is_a?(String)
      visit.call(child)
    end
  when Array
    value.each { |child| visit.call(child) }
  end
end
visit.call(manifest.reject { |key, _| key == "adapters" })

missing_adapters = referenced_adapters.uniq - declared_adapters.keys
missing_adapters.sort.each do |adapter|
  errors << "Referenced adapter '#{adapter}' is not declared under top-level 'adapters'"
end

contract_reference = manifest.dig("metadata", "contract")
if !contract_reference.is_a?(String) || contract_reference.empty?
  errors << "metadata.contract must be a non-empty local path"
else
  contract_path = manifest_path.dirname.join(contract_reference).cleanpath
  relative_contract = contract_path.relative_path_from(repo_root).to_s
  if relative_contract == ".." || relative_contract.start_with?("../")
    errors << "metadata.contract must resolve inside the repository"
  elsif !contract_path.file?
    errors << "metadata.contract does not exist: #{relative_contract}"
  end
end

if errors.any?
  errors.each { |error| warn "ERROR: #{error}" }
  exit 1
end

puts "Validated #{manifest_path.relative_path_from(repo_root)}"
puts "#{declared_adapters.length} adapters declared; #{referenced_adapters.uniq.length} referenced"
